text/0000-guaranteed-slice-repr.md

```diff
@@ -0,0 +1,182 @@
+- Feature Name: guaranteed_slice_repr
+- Start Date: 2025-02-18
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC guarantees the in-memory representation of slice and str references.
+Specifically, `&[T]` and `&mut [T]` are guaranteed to have the same layout as:
+
+```rust
+#[repr(C)]
+struct Slice<T> {
+    data: *const T,
+    len: usize,
```

Comment:
Why this specific order? Why not length before pointer? Are there any plausible reasons to prefer one over the other, e.g., based on target architecture?

Reply:
Note that our current layout algorithm is best able to exploit any niche in `len` if `len` is first in this ordering. And we currently cannot use a "double niche" on two different fields very effectively. I can work on improving this, but I cannot promise a specific result yet.

Due to various quirks of our existing APIs, the `len` niche would not apply to
1. zero-sized types
2. the "raw" version of this: `*mut [T]` and `*const [T]`

Reply:
@workingjubile Why would `len` have a niche? I would expect `data` to be the field with a niche, as it is non-null for `&[T]`.

Reply:
> The total size len * mem::size_of::<T>() of the slice must be no larger than isize::MAX, and adding that size to data must not “wrap around” the address space. See the safety documentation of [pointer::offset](https://doc.rust-lang.org/std/primitive.pointer.html#method.offset).

https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html

Reply:
Oh, clever! Thanks for describing that-- I'll mention it in the RFC.

Reply:
> as has been FCPed by opsem in [rust-lang/unsafe-code-guidelines#510](https://github.com/rust-lang/unsafe-code-guidelines/issues/510)

Theoretically, we could have been even stricter, additionally requiring that ptr + length not overflow the address space.

Reply:
I have a [draft PR](https://github.com/rust-lang/rust/pull/116542) where I tried implementing the niche in the length field, it didn't yield much perf improvements. Maybe I should try separating the metadata and the niche. And I haven't attempted reordering the fields yet either.